### PR TITLE
BUGFIX fix injection bhp pressure for CO2/H2 STORE

### PR DESCRIPTION
**Pull request by totto82**
The b factor was doubled up for the gas component if DISGASW and VAPWAT was activated. Should only affect the bhp calculations, not the reservoir results itself.

### CODE REVIEW

**Comment by totto82** [comment]
jenkins build this please

File: opm/simulators/wells/StandardWell_impl.hpp [hunk 1]
```diff
@@ -304,8 +304,10 @@ namespace Opm
                     volumeRatio += cmix_s[oilCompIdx] / b_perfcells_dense[oilCompIdx];
                 }
                 if (FluidSystem::phaseIsActive(FluidSystem::gasPhaseIdx)) {
-                    const unsigned gasCompIdx = Indices::canonicalToActiveComponentIndex(FluidSystem::gasCompIdx);
-                    volumeRatio += cmix_s[gasCompIdx] / b_perfcells_dense[gasCompIdx];
+                    if (! (FluidSystem::enableVaporizedWater() && FluidSystem::enableDissolvedGasInWater())) {
+                        const unsigned gasCompIdx = Indices::canonicalToActiveComponentIndex(FluidSystem::gasCompIdx);
```
Comment (GitPaean) [inline opm/simulators/wells/StandardWell_impl.hpp:308]: It is a good catch, but maybe you can put this in the else branch of this if condition above.  The logic here is rather complicated, that is why we had the fallout in the first place. 

```c++
            if (FluidSystem::enableVaporizedWater() && FluidSystem::enableDissolvedGasInWater()) {
                disOilVapWatVolumeRatio(volumeRatio, rvw, rsw, pressure,
                                        cmix_s, b_perfcells_dense, deferred_logger);
            } else {
```

**Comment by totto82** [comment]
The test failure is expected since bhp is changed. i.e. this will need an update of data after review.

File: opm/simulators/wells/StandardWell_impl.hpp (hunk 1 above)
Comment (totto82) [reply opm/simulators/wells/StandardWell_impl.hpp:308]: Seems like a good idea. I will try to improve the code logic.

**Comment by totto82** [comment]
jenkins build this please

File: opm/simulators/wells/StandardWell_impl.hpp [hunk 2]
```diff
@@ -279,36 +279,42 @@ namespace Opm
...
             } else {
+
                 if (FluidSystem::phaseIsActive(FluidSystem::waterPhaseIdx)) {
                     const unsigned waterCompIdx = Indices::canonicalToActiveComponentIndex(FluidSystem::waterCompIdx);
                     volumeRatio += cmix_s[waterCompIdx] / b_perfcells_dense[waterCompIdx];
                 }
+
+                if (FluidSystem::phaseIsActive(FluidSystem::oilPhaseIdx) && FluidSystem::phaseIsActive(FluidSystem::gasPhaseIdx)) {
+                    gasOilVolumeRatio(volumeRatio, rv, rs, pressure,
+                                    cmix_s, b_perfcells_dense, deferred_logger);
```
Comment (akva2) [inline opm/simulators/wells/StandardWell_impl.hpp:299]: meh

File: opm/simulators/wells/StandardWell_impl.hpp [hunk 3]
```diff
@@ -279,36 +279,42 @@ namespace Opm
...
+                } else {
+                    if (FluidSystem::phaseIsActive(FluidSystem::oilPhaseIdx)) {
+                        const unsigned oilCompIdx = Indices::canonicalToActiveComponentIndex(FluidSystem::oilCompIdx);
+                        volumeRatio += cmix_s[oilCompIdx] / b_perfcells_dense[oilCompIdx];
+                    }
+                    if (FluidSystem::phaseIsActive(FluidSystem::gasPhaseIdx)) {
+                        if (! (FluidSystem::enableVaporizedWater() && FluidSystem::enableDissolvedGasInWater())) {
```
Comment (GitPaean) [inline opm/simulators/wells/StandardWell_impl.hpp:306]: `if (! (FluidSystem::enableVaporizedWater() && FluidSystem::enableDissolvedGasInWater())) ` is not needed here.

File: opm/simulators/wells/StandardWell_impl.hpp [hunk 4]
```diff
@@ -279,36 +279,42 @@ namespace Opm
 
             // compute volume ratio between connection at standard conditions
             Value volumeRatio = bhp * 0.0; // initialize it with the correct type
-;
+
             if (FluidSystem::enableVaporizedWater() && FluidSystem::enableDissolvedGasInWater()) {
                 disOilVapWatVolumeRatio(volumeRatio, rvw, rsw, pressure,
                                         cmix_s, b_perfcells_dense, deferred_logger);
+                // only supported for gas-water case
```
Comment (GitPaean) [inline opm/simulators/wells/StandardWell_impl.hpp:286]: I leave @bska and @akva2  to comment whether we want to use `assert` here.  

It basically means that we enter a simulation not allowed at the moment, which is have water-gas mixing while we might not have water or gas, or we have oil active.  We do not support that.

Comment (akva2) [reply opm/simulators/wells/StandardWell_impl.hpp:286]: even though we ideally would have failed earlier, using assert for things that can be triggered by user input (files) is never the correct solution.

Comment (totto82) [reply opm/simulators/wells/StandardWell_impl.hpp:286]: The Simulator will throw in the setup face. I added the asserts for the programmer if/when we start thinking about adding disgasw for a blackoil run. I can remove them and add a throw instead.

Comment (akva2) [reply opm/simulators/wells/StandardWell_impl.hpp:286]: if it throws earlier it's all good.

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
All comments are addressed. @GitPaean Can I start the process of updating the data and get this in?

**Review by GitPaean — APPROVED** [review]
Thanks for the update. The changes introduced in this PR look fine now.

File: opm/simulators/wells/StandardWell_impl.hpp [hunk 5]
```diff
@@ -279,33 +279,39 @@ namespace Opm
 
             // compute volume ratio between connection at standard conditions
             Value volumeRatio = bhp * 0.0; // initialize it with the correct type
-;
+
             if (FluidSystem::enableVaporizedWater() && FluidSystem::enableDissolvedGasInWater()) {
```
Comment (GitPaean) [inline opm/simulators/wells/StandardWell_impl.hpp:283]: This is not in this PR, so I approved the PR. 

A discussion here, should here be 
```c++
if (FluidSystem::enableVaporizedWater() || FluidSystem::enableDissolvedGasInWater()) {
```
??

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
The regression looks mostly fine. The well INJ looks like injecting with constant gas rate and there is some bhp/pressure change.

**Comment by GitPaean** [comment]
jenkins build this opm-tests=1228 please

**Pull request merged**